docs: update code in K8s tutorial, with source in repo (part 2)

dwilding opened the pull request:
This PR adds the remaining example charms from the Kubernetes charm tutorial, continuing from #1718.

At the same time, I did a fair bit of cleanup: linting, formatting, replacing links to juju.is, making comments/docstrings consistent, and fixing small mistakes in the docs.

One significant change to the tutorial is that I've switched to using `charmcraft fetch-libs`, with libs specified in charmcraft.yaml.

To help with reviewing the changes, I've added comments at points that are particularly significant.

dwilding commented on docs/tutorial/from-zero-to-hero-write-your-first-kubernetes-charm/integrate-your-charm-with-postgresql.md:
````diff
@@ -402,41 +412,51 @@ In this chapter, we also defined a new method `_on_collect_status` that checks v
+    assert state_out.unit_status == testing.ActiveStatus()
+```
+
+Since `test_pebble_layer` doesn't arrange a database relation, the unit will be in `blocked` status instead of `active`.
````
So why don't we change the unit test to assert this, instead of removing the assertion?

dwilding commented on examples/k8s-3-postgresql/tests/unit/test_charm.py:
```diff
@@ -0,0 +1,125 @@
+    )
+
+
+def test_relation_data():
```
This unit test produces a warning:

```
tests/unit/test_charm.py::test_relation_data
  /home/ubuntu/fastapi-demo/lib/charms/data_platform_libs/v0/data_interfaces.py:998: DeprecationWarning: JujuVersion.from_environ() is deprecated, use self.model.juju_version instead
    self._jujuversion = JujuVersion.from_environ()
```

Is there anything we should do about that?

benhoyt replied:
Please open PR on their repo https://github.com/canonical/data-platform-libs/

benhoyt commented:
@dwilding Per discussion, please:

1) Ensure examples/ doesn't end up in sdist build
2) Remove the charm lib from the repo, and run `charmcraft fetch-libs` before packing (in tests and in the doc)

dwilding commented:
>     1. Ensure examples/ doesn't end up in sdist build

I ran `python -m build --sdist` and confirmed that `examples` isn't added to the build.
 
>     2. Remove the charm lib from the repo, and run `charmcraft fetch-libs` before packing (in tests and in the doc)

I removed the lib and updated the "Example Charm Tests" workflow to run `charmcraft fetch-lib` before `tox -e unit`.

This PR won't change the instructions in the docs about how to check out the tutorial code. I'll do that in #1738.

benhoyt commented on .github/workflows/example-charm-tests.yaml:
```diff
@@ -30,6 +34,13 @@ jobs:
+      - name: Fetch any charmlibs
+        run: |
+          cd ${{ matrix.dir }}
+          # Gets the job done... Better for tox.ini to have a dedicated env for fetching charmlibs?
```
https://canonical-charmcraft.readthedocs-hosted.com/stable/reference/files/charmcraft-yaml-file/#charm-libs

dwilding commented on docs/tutorial/from-zero-to-hero-write-your-first-kubernetes-charm/observe-your-charm-with-cos-lite.md:
```diff
@@ -27,31 +27,46 @@ In the charming universe, what you would do is deploy the existing [Canonical Ob
-Follow the steps below to make your charm capable of integrating with the existing [Prometheus](https://charmhub.io/prometheus-k8s) charm. This will enable your charm user to collect real-time metrics about your application.
-
-### Fetch the Prometheus interface libraries
+## Fetch libraries
```
This doc had several almost identical sections where we asked the charmer to download libs individually. I've replaced those sections by a new section that asks the charmer to specify all the libs in charmcraft.yaml, then use `charmcraft fetch-libs`. This helps our CI and is a much nicer experience for the charmer!

dwilding commented on examples/README.md:
```diff
@@ -1,12 +1,22 @@
-This directory contains charms that are used as examples in the Ops documentation. **It's currently a work in progress.**
+This directory contains charms that you can pack and deploy locally, to help learn how to use Ops.
```
This file is the new "landing page" for our example charms. Here's what it looks like:
https://github.com/dwilding/canonical-operator/tree/add-tutorial-charm-2/examples

dwilding commented on pyproject.toml:
```diff
@@ -139,6 +139,9 @@ extend-exclude = ["docs/conf.py", "docs/.sphinx/"]
 # Ruff formatter configuration
 [tool.ruff.format]
 quote-style = "single"
+exclude = [
+    "examples/*/lib/*",  # Ignore libs fetched during our testing (or by a charmer)
+]
```
We don't include libs in the Ops repo. But if someone clones the repo and downloads any libs, they'll need to be excluded from Ops linting and formatting.

tonyandrewmeyer commented on docs/tutorial/from-zero-to-hero-write-your-first-kubernetes-charm/integrate-your-charm-with-postgresql.md:
```diff
@@ -252,20 +262,20 @@ And define a method that does the various checks, adding appropriate statuses. T
     port = self.config['server-port']
     if port == 22:
-        event.add_status(ops.BlockedStatus('Invalid port number, 22 is reserved for SSH'))
+        event.add_status(ops.BlockedStatus('invalid port number, 22 is reserved for SSH'))
```
Are these to match the code, or is our recommendation that the message starts with a lower case? I can't remember if we have an official recommendation on this or not, and couldn't find one with a quick search. Probably we should opinionated on this, and have that opinion in the API reference for the status classes?

james-garner-canonical replied:
+1 on Tony's comment. If this is just making the docs and code match, I'm more than happy for the code to be the one changed in this case.

dwilding replied:
If I remember correctly, I think the docs were inconsistent, so I switched everything to lowercase since that's what Juju seems to output. I also can't find a specific recommendation. But the code in https://github.com/canonical/juju-sdk-tutorial-k8s uses uppercase, so I'll stick with uppercase for now

tonyandrewmeyer commented on examples/k8s-2-configurable/charmcraft.yaml:
```diff
@@ -0,0 +1,43 @@
+  demo-fastapi-k8s
+description: |
+  This is a demo charm built on top of a small Python FastAPI server.
+  This charm can be integrated with the PostgreSQL charm and COS Lite bundle (Canonical Observability Stack).
```
Not true at this point, but maybe it's simpler to just keep the description consistent? I certainly don't think we would want to have in the tutorial contents instructions to keep updating this field!

james-garner-canonical replied:
I'm fine with leaving the description as is, but another idea might be to drop the second line altogether since that information is ultimately exposed in `provides` and `requires`?

dwilding replied:
I like James's idea - I've gone with that

tonyandrewmeyer commented on examples/k8s-5-observe/charmcraft.yaml:
```diff
@@ -0,0 +1,84 @@
+
+actions:
+  get-db-info:
+    description: Fetches Database authentication information
```
```suggestion
    description: Fetches database authentication information
```

tonyandrewmeyer commented on examples/k8s-4-action/charmcraft.yaml:
```diff
@@ -0,0 +1,61 @@
+
+actions:
+  get-db-info:
+    description: Fetches Database authentication information
```
```suggestion
    description: Fetches database authentication information
```

tonyandrewmeyer commented on examples/k8s-4-action/charmcraft.yaml:
```diff
@@ -0,0 +1,61 @@
+      type: int
+
+requires:
+  database:
+    interface: postgresql_client
+    limit: 1
```
Feel free to say that we should leave this for a follow-up PR: we now strongly recommend that the `optional` field is always included for all the provides/requires relations, rather than relying on the default.
```suggestion
  database:
    interface: postgresql_client
    limit: 1
    optional: false
```

tonyandrewmeyer commented on examples/k8s-3-postgresql/charmcraft.yaml:
```diff
@@ -0,0 +1,52 @@
+      type: int
+
+requires:
+  database:
+    interface: postgresql_client
+    limit: 1
```
```suggestion
  database:
    interface: postgresql_client
    limit: 1
    optional: false
```

tonyandrewmeyer commented on examples/k8s-5-observe/charmcraft.yaml:
```diff
@@ -0,0 +1,84 @@
+      description: Default port on which FastAPI is available
+      type: int
+
+requires:
+  database:
+    interface: postgresql_client
+    limit: 1
+  log-proxy:
+    interface: loki_push_api
+    limit: 1
+
+provides:
+  metrics-endpoint:
+    interface: prometheus_scrape
+  grafana-dashboard:
+    interface: grafana_dashboard
```
```suggestion
requires:
  database:
    interface: postgresql_client
    limit: 1
    optional: false
  log-proxy:
    interface: loki_push_api
    limit: 1
    optional: true

provides:
  metrics-endpoint:
    interface: prometheus_scrape
    optional: true
  grafana-dashboard:
    interface: grafana_dashboard
    optional: true
```

tonyandrewmeyer commented on examples/k8s-5-observe/tests/integration/test_charm.py:
```diff
@@ -0,0 +1,65 @@
+    await ops_test.model.integrate(f'{APP_NAME}', 'postgresql-k8s')
+    await ops_test.model.wait_for_idle(
+        apps=[APP_NAME], status='active', raise_on_blocked=False, timeout=300
+    )
```
It shouldn't be in this PR, but we should add tests that verify the other possible relations (which is what we recommend as baseline integration tests). Do we have that tracked somewhere?

james-garner-canonical commented on docs/tutorial/from-zero-to-hero-write-your-first-kubernetes-charm/make-your-charm-configurable.md:
```diff
@@ -191,30 +190,36 @@ First, we'll add a test that sets the port in the input state and asserts that t
     )
     state_out = ctx.run(ctx.on.config_changed(), state_in)
-    assert "--port=8080" in state_out.get_container(container.name).layers["fastapi_demo"].services["fastapi-service"].command
+    assert (
+        '--port=8080'
+        in state_out.get_container(container.name)
+        .layers['fastapi_demo']
+        .services['fastapi-service']
+        .command
+    )
```
Oh wow I haven't seen this formatting before. How about this:
```suggestion
    command = (
        state_out.get_container(container.name)
        .layers["fastapi_demo"]
        .services["fastapi-service"]
        .command
    )
    assert "--port=8080" in command
```

dwilding replied:
I've gone with your suggestion, thanks!

james-garner-canonical commented on examples/k8s-2-configurable/tox.ini:
```diff
@@ -0,0 +1,66 @@
+    ops[testing]
+    -r {tox_root}/requirements.txt
+commands =
+    coverage run --source={[vars]src_path} \
+                 -m pytest \
+                 --tb native \
+                 -v \
+                 -s \
+                 {posargs} \
+                 {[vars]tests_path}/unit
```
Feel free to leave this as is, since I guess we might update these files anyway to use `uv`, but I really prefer formatting like this:
```suggestion
    coverage run --source={[vars]src_path} -m pytest \
        -v \
        -s \
        --tb native \
        {posargs} \
        {[vars]tests_path}/unit
```
That is, the first line is our `pytest` invocation (via `coverage`), and the `pytest` args are just indented 4 spaces from that. The argument lines can then be identical for the `integration` environment's `pytest` command (+/- `--log-cli-level=INFO`, and with the test directory differing).

dwilding replied:
Nice tip, thanks! I've just done this.

Are you also suggesting that I change the `integration` environments?

```
commands =
    pytest \
        -v \
        -s \
        --tb native \
        --log-cli-level=INFO \
        {posargs} \
        {[vars]tests_path}/integration
```

dwilding replied:
Done!

james-garner-canonical reviewed (approved):
Thanks for doing this! No major blockers on merging this imo, though it would be good to resolve the `ops` version specifier question.

dwilding commented:
I've verified that integration tests pass for each of the charms. Merging now. Thanks again!
